repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py

```diff
@@ -0,0 +1,670 @@
+            conf += "maxdistance {}\n".format(maxdistance)
+            conf += "\n"
+
+        conf += "# Increase the minimum number of selectable sources required to adjust\n"
+        conf += "# the system clock.\n"
+        if minsources > 1:
+            conf += "minsources {}\n".format(minsources)
+        else:
+            conf += "#minsources 2\n"
+        conf += "\n"
+
+        conf += self.get_chrony_conf_allows()
+
+        conf += self.get_chrony_conf_cmdallows()
+
+        conf += "# Serve time even if not synchronized to a time source.\n"
+        if orphan_stratum > 0 and orphan_stratum < 16:
+            conf += "local stratum {} orphan\n".format(orphan_stratum)
+        elif local_stratum > 0 and local_stratum < 16:
+            conf += "local stratum {}\n".format(local_stratum)
+        else:
+            conf += "#local stratum 10\n"
+        conf += "\n"
+
+        conf += "# Specify file containing keys for NTP authentication.\n"
+        conf += "keyfile {}\n".format(chrony_keys_path)
+        conf += "\n"
+
+        conf += "# Get TAI-UTC offset and leap seconds from the system tz database.\n"
+        conf += "leapsectz right/UTC\n"
+        conf += "\n"
+
+        conf += "# Specify directory for log files.\n"
+        conf += "logdir /var/log/chrony\n"
+        conf += "\n"
+
+        conf += "# Select which information is logged.\n"
+        if logs:
+            conf += "log {}\n".format(" ".join(logs))
+        else:
+            conf += "#log measurements statistics tracking\n"
+
+        return conf
+
+    def get_chrony_keys(self):
+        if not self.keyfile:
+            return ""
+
+        keys = "# This file was converted from {}.\n".format(self.keyfile)
+        keys += "\n"
+
+        for key in self.keys:
+            id = key[0]
+            type = key[1]
+            password = key[2]
+
+            if type in ["m", "M"]:
+                type = "MD5"
+            elif type not in ["MD5", "SHA1", "SHA256", "SHA384", "SHA512"]:
+                continue
+
+            prefix = "ASCII" if len(password) <= 20 else "HEX"
+
+            for first, last in self.trusted_keys:
+                if first <= id <= last:
+                    trusted = True
+                    break
+            else:
+                trusted = False
+
+            # Disable keys that were not marked as trusted
+            if not trusted:
+                keys += "#"
+
+            keys += "{} {} {}:{}\n".format(id, type, prefix, password)
+
+        return keys
+
+    def write_file(self, path, mode, content, backup):
+        path = self.root_dir + path
+        if backup and os.path.isfile(path):
+            os.rename(path, path + ".old")
+
+        with open(os.open(path, os.O_CREAT | os.O_WRONLY | os.O_EXCL, mode), "w",
+                  encoding=self.file_encoding) as f:
+            logging.info("Writing %s", path)
+            f.write(u"" + content)
```

Comment:
f.write(unicode(content))?

Reply:
Is there a difference?

Reply:
As in unnecessary string concatenation with forceful unicode convertion vs direct unicode convertion